### PR TITLE
Eagerly populate the class dict for cudf.pandas proxy types

### PR DESCRIPTION
**Pull request by shwina**
Rather than dynamically looking up class attributes (and methods), this PR makes it so that we eagerly populate the class with all known methods and attributes (by inspecting the "slow" class).

This solves a number of problems:

- it makes `getattr` trivially inexpensive (no dynamic `__getattr__` for each attribute access)
- it ensures the _same_ object is returned every time you do, e.g., `DataFrame.max`
- it makes tab completion fast because the attributes don't have to be computed each time
- it no longer exposes attributes that are specific to cuDF - for example `Series.list`
- it allows subclassing of proxy types to work better. For example, derived types can now call `super().` to access attributes of base types

### CODE REVIEW

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (shwina) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:469]: Note - there's no longer dynamic attribute lookup via `__getattr__`.

File: python/cudf/cudf/pandas/_wrappers/pandas.py
Comment (bdice) [inline python/cudf/cudf/pandas/_wrappers/pandas.py:98]: TIL a new part of the descriptors object model!

https://docs.python.org/3/howto/descriptor.html#closing-thoughts

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (bdice) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:788]: These just reassign None when the value is None, but in a slightly nonobvious way. Let’s simplify to
```python
if assigned is None:
    assigned = functools.WRAPPER_ASSIGNMENTS
```
and similarly for `updated`.

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (bdice) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:895]: Generally use the “normal” ways of accessing the object model even if it’s a part of defining the object model. Same for things like `__hash__` and `__eq__` and `__repr__`. (If that makes sense…)
```suggestion
        return dir(self._fsproxy_slow)
```

**Review by bdice — APPROVED** [review]

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (mroeschke) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:224]: For arithmetic dunder methods like `__add__`, wouldn't we want to create a `_FastSlowAttribute` for that?

Comment (shwina) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:224]: So those are defined in the base class `_FastSlowProxy` now concretely (so it's already in the class dictionary).

**Comment by shwina** [comment]
> Generally looks good to me. Any impact on performance or the pass rate of the pandas test suite?

w.r.t benchmarks (and assuming we can unblock them) - would an improvement (or at least no degradation) in the time it takes to run the pandas unit tests be a good benchmark? (@bdice)

File: python/cudf/cudf/pandas/_wrappers/pandas.py
Comment (mroeschke) [inline python/cudf/cudf/pandas/_wrappers/pandas.py:752]: Curious, does this _need_ to be the exact location of `eval` instead of using the public `pd.eval`?

Similar to the register functions below e.g. `pd.api.extensions.register_dataframe_accessor`?

Comment (shwina) [reply python/cudf/cudf/pandas/_wrappers/pandas.py:752]: I think that used to be the case when we first wrote it, but no longer so

Comment (mroeschke) [reply python/cudf/cudf/pandas/_wrappers/pandas.py:752]: Ah OK. Would be nice if this could be avoided but not a big deal

**Comment by galipremsagar** [comment]
@vyasr It looks like there are hangs, I'm currently debugging them locally. The run is currently taking more than 2 hours. I'll push the changes to this PR once I have them ready.

**Comment by vyasr** [comment]
Got it, thanks for the update @galipremsagar! I was clearing out my backlog and thought this PR might be ready.

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (wence-) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:530]: *question*: these changes are because `operator.add` might do `y.__radd__(x)`, which we don't want?

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (wence-) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:1060]: Debugging?

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (wence-) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:1064]: Debugging?

**Review by wence- — COMMENTED** [review]
*question*: It seems like this branch is also pulling in a bunch of unrelated (?) changes in cudf-proper. Is that deliberate in the sense these are bug-fixes for issues that are exposed by these changes?

**Comment by galipremsagar** [comment]
> _question_: It seems like this branch is also pulling in a bunch of unrelated (?) changes in cudf-proper. Is that deliberate in the sense these are bug-fixes for issues that are exposed by these changes?

Yes, those are all the bug-fixes for the issues that are exposed by these changes, failures can be found here: https://github.com/rapidsai/cudf/actions/runs/8916717301/attempts/2#summary-24512387065

**Comment by wence-** [comment]
> > _question_: It seems like this branch is also pulling in a bunch of unrelated (?) changes in cudf-proper. Is that deliberate in the sense these are bug-fixes for issues that are exposed by these changes?
> 
> Yes, those are all the bug-fixes for the issues that are exposed by these changes, failures can be found here: https://github.com/rapidsai/cudf/actions/runs/8916717301/attempts/2#summary-24512387065
> 
> The initial set of changes caused 12% drop in pass-rate, with the current bug fixes we are down to 2% decrease (still WIP)

Aha, thanks!

File: python/cudf/cudf/pandas/_wrappers/pandas.py
Comment (wence-) [inline python/cudf/cudf/pandas/_wrappers/pandas.py:1310]:
```suggestion
    "EasterMonday",
```

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:548]: Yes, I forked out these changes to a separate PR: https://github.com/rapidsai/cudf/pull/15684

File: ci/cudf_pandas_scripts/pandas-tests/run.sh
Comment (galipremsagar) [inline ci/cudf_pandas_scripts/pandas-tests/run.sh:25]:
```suggestion
```

**Review by galipremsagar — APPROVED** [review]

**Comment by galipremsagar** [comment]
@wence- I've reverted the vast majority of the changes I made to this PR and this is now ready for review. Since I've made a couple of minor fixes to this pr I'd like to get it reviewed by someone other than me before merging.

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:858]: I'm not sure why, but using `dir` instead of `.__dir__()` is causing 500 more failures consistently, in the interest of time I haven't dug into those failures.

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (mroeschke) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:785]:
```suggestion
    def __init__(self, name: str, *, private: bool = False):
```

Would be nice to make `private` keyword only

File: python/cudf/cudf/pandas/_wrappers/numpy.py
Comment (mroeschke) [inline python/cudf/cudf/pandas/_wrappers/numpy.py:126]:
```suggestion
        "base": _FastSlowAttribute("base", private=True),
```

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/fast_slow_proxy.py]: Done 👍

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (mroeschke) [inline python/cudf/cudf/pandas/fast_slow_proxy.py:809]: Would be good to adjust the `func` typing in _maybe_wrap_result as `func: Callable | None` and raise an exception in `is_final_type` and `is_intermediate_type` branches `if func is None`

File: python/cudf/cudf/pandas/_wrappers/pandas.py
Comment (mroeschke) [inline python/cudf/cudf/pandas/_wrappers/pandas.py:281]: So these attributes we are wanting the fast path to have a chance to evaluate and not automatically be `_Unusable()`?

Comment (galipremsagar) [reply python/cudf/cudf/pandas/_wrappers/pandas.py:281]: yes

File: python/cudf/cudf/pandas/scripts/run-pandas-tests.sh
Comment (mroeschke) [inline python/cudf/cudf/pandas/scripts/run-pandas-tests.sh:141]: So these are still confirmed failing with now eagerly populating the class dict?

File: python/cudf/cudf/pandas/_wrappers/pandas.py
Comment (mroeschke) [reply python/cudf/cudf/pandas/_wrappers/pandas.py:281]: Ah so in this case since it's `private` we won't try to operate on e.g. `cudf.Index._data` which is a `ColumnAccessor` and instead reach for `cudf.Index._data` which is a numpy ndarray which we can wrap.

This part of the eager population I'm not too thrilled about. These private variables are not guaranteed to be stable. It would be nice if any slow attribute was attempted to return a proxy object so we didn't have to specify attributes like this but I suppose that can be looked at in a follow up

File: python/cudf/cudf/pandas/scripts/run-pandas-tests.sh
Comment (galipremsagar) [reply python/cudf/cudf/pandas/scripts/run-pandas-tests.sh:141]: yes, some of these segfaults and a few take quite a long time. The plan is to tackle each of them separately as follow-ups.

File: python/cudf/cudf/pandas/_wrappers/pandas.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/_wrappers/pandas.py:281]: Yeah, there are still such cases we can address in follow-ups. At this point the scope of this PR has kept on expanding, that's why I haven't addressed additional issues in this PR.

**Review by mroeschke — APPROVED** [review]
My typing request and concern about having to define private attributes in our proxies can be addressed in a follow up

File: python/cudf/cudf/pandas/fast_slow_proxy.py
Comment (galipremsagar) [reply python/cudf/cudf/pandas/fast_slow_proxy.py:809]: This doesn't seem to be straight-forward, while testing this change I noticed `4k` additional failures. Maybe we can revisit this in another PR.

**Pull request merged**
